### ISSUE TITLE
Memory leak when using a single connection pool intensively

### ISSUE THREAD

**Issue by tsergeyev** [akka/akka-http#1637]
Making lots of requests using a single connection pool consumes memory and never releases it back:

![connectionpool](https://user-images.githubusercontent.com/34548790/34003173-aa04feb4-e105-11e7-906d-8c5c656ea507.png)

When using connection-level API with no pooling, everything is perfectly fine:

![single connection](https://user-images.githubusercontent.com/34548790/34003631-d4511ecc-e106-11e7-81d4-c18c70fc7e1e.png)

Standard host-level API was used in the first scenario (with memory leak):

```scala
Http().singleRequest(request)
```

Standard connection-level API was used in the second scenaro (no memory leak):

```scala
val connectionFlow = Http().outgoingConnection(request.uri.authority.host.address, request.uri.effectivePort)
Source.single(request).via(connectionFlow).runWith(Sink.head)
```

After checking allocations, it seems that lots of streams are being created (all items in red are steadily growing, except for a few ones):

![allocations](https://user-images.githubusercontent.com/34548790/34003676-f8e3f1c4-e106-11e7-93ea-6cc1d8509c2d.png)

Below is the code used to produce these charts. It creates 50 actors hammering the same server with some delay between requests. Setting `TestApp.useConnectionPool` to `true` results in `OutOfMemoryError`  in less than 2 minutes, while setting it to `false` makes the test run infinitely with no issues.

```scala
import akka.actor.{Actor, ActorLogging, ActorSystem, Props}
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Sink, Source}

import scala.concurrent.duration._
import scala.concurrent.{ExecutionContext, Future}
import scala.util.Random

class HttpClient(baseUri: String, contentLength: Int, useConnectionPool: Boolean) extends Actor with ActorLogging {
  implicit val system: ActorSystem = context.system
  implicit val executionContext: ExecutionContext = context.dispatcher
  implicit val materializer: ActorMaterializer = ActorMaterializer()

  case object Download

  val delayMillis = 1500
  val delayJitter = 1000

  scheduleDownload()

  override def receive = {
    case Download =>
      fireRequest(HttpRequest(uri = baseUri + "/?size=" + Random.nextInt(contentLength))).map { httpResponse =>
        Unmarshal(httpResponse.entity).to[String].map { content =>
          log.info("Received " + content.length + " bytes: " + httpResponse.status)
          scheduleDownload()
        }.recover {
          case throwable =>
            log.error("Error: " + throwable)
            httpResponse.discardEntityBytes()
        }
      }
  }

  def scheduleDownload(): Unit = {
    context.system.scheduler.scheduleOnce((delayMillis - delayJitter / 2 + Random.nextInt(delayJitter)).millis, self, Download)
  }

  private def fireRequest(request: HttpRequest): Future[HttpResponse] = {
    if (useConnectionPool) {
      Http().singleRequest(request)
    } else {
      val connectionFlow = Http().outgoingConnection(request.uri.authority.host.address, request.uri.effectivePort)
      Source.single(request).via(connectionFlow).runWith(Sink.head)
    }
  }
}

object TestApp extends App {
  val useConnectionPool = true // Setting this to true results in memory leak
  val contentLength = 256 * 1024 // Increasing this results in OutOfMemory happening sooner

  val system: ActorSystem = ActorSystem("test")

  for (i <- 1 to 50) {
    system.actorOf(Props[HttpClient](new HttpClient("http://localhost:9999", contentLength, useConnectionPool)))
  }
}

```
Here's the server part too (I prefer to run it separately to isolate client-side issues):

```scala
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.server.Directives._
import akka.stream.ActorMaterializer

object HttpServer extends App {
  implicit val system: ActorSystem = ActorSystem("test")
  implicit val materializer: ActorMaterializer = ActorMaterializer()

  val route =
    parameters('size.as[Int]) { size =>
      get {
        complete("#" * size)
      }
    }
  Http().bindAndHandle(route, "0.0.0.0", 9999)
}
```
Two final notes:
- Memory seems to be released at least partially when connection pool times out, but this is not proven.
- New connection pool behaves in the same way as the legacy connection pool (checked with 10.0.11).

**Comment by jrudolph** [comment]
Thanks for the report, @tsergeyev. Which Akka version are you running on? There was a problem on Akka 2.5.7 that could lead to a memory leak that is fixed when running with Akka 2.5.8. See https://github.com/akka/akka-http/issues/1613 for more info.

**Comment by tsergeyev** [comment]
Thank you for quick reply! You were right, everything looks great after updating to Akka 2.5.8:

![connectionpool - fixed](https://user-images.githubusercontent.com/34548790/34035022-b27252ae-e191-11e7-982b-f4a436c4dc85.png)

Closing the issue. Thanks!
